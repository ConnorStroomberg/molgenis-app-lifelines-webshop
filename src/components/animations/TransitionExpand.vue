<template>
  <transition
    name="expand"
    @enter="enter"
    @after-enter="afterEnter"
    @leave="leave"
  >
    <slot/>
  </transition>
</template>

<script>
// https://github.com/maoberlehner/transition-to-height-auto-with-vue
// https://markus.oberlehner.net/blog/transition-to-height-auto-with-vue/

export default {
  name: `TransitionExpand`,
  methods: {
    afterEnter (element) {
      element.style.height = 'auto'
    },
    enter (element) {
      const { width } = getComputedStyle(element)

      element.style.width = width
      element.style.position = 'absolute'
      element.style.visibility = 'hidden'
      element.style.height = 'auto'

      const { height } = getComputedStyle(element)

      element.style.width = null
      element.style.position = null
      element.style.visibility = null
      element.style.height = 0

      // Force repaint to make sure the
      // animation is triggered correctly.
      // eslint-disable-next-line no-unused-expressions
      getComputedStyle(element).height

      setTimeout(() => {
        element.style.height = height
      })
    },
    leave (element) {
      const { height } = getComputedStyle(element)

      element.style.height = height

      // Force repaint to make sure the
      // animation is triggered correctly.
      // eslint-disable-next-line no-unused-expressions
      getComputedStyle(element).height

      setTimeout(() => {
        element.style.height = 0
      })
    }
  }
}
</script>

<style scoped>
  * {
    will-change: height;
    transform: translateZ(0);
    backface-visibility: hidden;
    perspective: 1000px;
  }

  .expand-enter-active,
  .expand-leave-active {
    transition: height 0.2s ease-in-out;
    overflow: hidden;
  }

  .expand-enter,
  .expand-leave-to {
    height: 0;
  }
</style>
