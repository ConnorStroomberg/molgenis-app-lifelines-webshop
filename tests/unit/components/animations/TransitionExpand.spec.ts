import { mount, Wrapper } from '@vue/test-utils'
import Vue from 'vue'
import TransitionExpand from '@/components/animations/TransitionExpand.vue'

describe('TransitionExpand.vue', () => {
  let wrapper: Wrapper<Vue>

  beforeEach(() => {
    wrapper = mount(TransitionExpand, {
    })
  })

  it('', () => {
    expect(wrapper).toMatchSnapshot()
  })
})
