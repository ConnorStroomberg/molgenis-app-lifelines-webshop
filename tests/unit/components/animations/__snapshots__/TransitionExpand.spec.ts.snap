// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionExpand.vue  1`] = ``;
